### PR TITLE
Add median implementation

### LINKED ISSUE #498: Implement support for various numpy reduce functions
Implement support for various numpy reduce functions like sum, mean, median, etc.

### PR DESCRIPTION
**Pull request by hargup**
fixes #498

**TODO**
- [x] Don't mutate the array for the computation of median of three.
- [x] Other Comments about the implementation.
- [x] Add typing information
- [x] Test for median_global
- [x] Tests for special cases
    - [x] sorted
    - [x] reverse sorted
    - [x] array with repeated values

### CODE REVIEW

**Comment by hargup** [comment]
The implementation of partition is directly taken from CLRS.

File: numba/tests/test_array_reductions.py [hunk 1]
```diff
@@ -148,6 +154,18 @@ def test_argmax_basic(self):
         npr, nbr = run_comparative(array_argmax, arr)
         self.assertPreciseEqual(npr, nbr)
 
+    def test_median_odd(self):
+        arr = np.arange(101)
+        np.random.shuffle(arr)
+        npr, nbr = run_comparative(array_argmax, arr)
```
Comment (pitrou) [inline numba/tests/test_array_reductions.py]: Apparently you are not testing the right function :)

File: numba/tests/test_array_reductions.py [hunk 2]
```diff
@@ -148,6 +154,18 @@ def test_argmax_basic(self):
...
+        self.assertPreciseEqual(npr, nbr)
+
+    def test_median_even(self):
+        arr = np.arange(100)
+        np.random.shuffle(arr)
+        npr, nbr = run_comparative(array_argmax, arr)
```
Comment (pitrou) [inline numba/tests/test_array_reductions.py]: Ditto.

File: numba/targets/arrayobj.py [hunk 3]
```diff
@@ -905,6 +905,44 @@ def array_argmax_impl(arry):
     return context.compile_internal(builder, array_argmax_impl, sig, args)
 
 
+@builtin
+@implement(numpy.median, types.Kind(types.Array))
+@implement("array.mean", types.Kind(types.Array))
```
Comment (pitrou) [inline numba/targets/arrayobj.py]: "array.median", rather.

File: numba/targets/arrayobj.py [hunk 4]
```diff
@@ -905,6 +905,44 @@ def array_argmax_impl(arry):
...
+def array_median(context, builder, sig, args):
+    # XXX: the implementation doesn't work for the case if array is not one
+    # dimentional
```
Comment (pitrou) [inline numba/targets/arrayobj.py]: s/dimentional/dimensional/

File: numba/targets/arrayobj.py [hunk 5]
```diff
@@ -905,6 +905,44 @@ def array_argmax_impl(arry):
...
+    def partition(A, p, r):
+        # TODO: make it randomized
+        x = A[r]
+        i = p-1
```
Comment (pitrou) [inline numba/targets/arrayobj.py]: To make things clearer, you could start at `i = p` and move `i += 1` after the swap.

File: numba/targets/arrayobj.py [hunk 6]
```diff
@@ -905,6 +905,44 @@ def array_argmax_impl(arry):
...
 
+@builtin
+@implement(numpy.median, types.Kind(types.Array))
+@implement("array.mean", types.Kind(types.Array))
+def array_median(context, builder, sig, args):
+    # XXX: the implementation doesn't work for the case if array is not one
+    # dimentional
+    def partition(A, p, r):
+        # TODO: make it randomized
+        x = A[r]
```
Comment (pitrou) [inline numba/targets/arrayobj.py]: This will make the algorithm O(n**2) if the array is already sorted. You could first swap `A[r]` and `A[(p + r) // 2]` to avoid that.

File: numba/targets/arrayobj.py [hunk 7]
```diff
@@ -905,6 +905,44 @@ def array_argmax_impl(arry):
...
+
+    def select(arry, k):
+        n = arry.shape[0]
+        temp_arry = arry.copy()
+        i = partition(temp_arry, 0, n-1)
+        while i != k:
+            if i < k:
+                i = partition(temp_arry, i+1, n-1)
+            else:
+                i = partition(temp_arry, 0, i-1)
```
Comment (pitrou) [inline numba/targets/arrayobj.py]: If this is not the first iteration, it seems this is doing too much here.
For example: n = 15, k = 7.
- first partition call: i = partition(0, 14) -> 3
- second partition call: i = partition(4, 14) -> 12
- third partition call: partition(0, 11)?? you should be calling partition(4, 11)...

File: numba/tests/test_array_reductions.py (hunk 1 above)
Comment (hargup) [inline numba/tests/test_array_reductions.py]: It seems I was really lousy with this PR, sorry will be more careful from
next time.

On 22 June 2015 at 19:48, Antoine Pitrou <notifications@github.com> wrote:

> In numba/tests/test_array_reductions.py
> <https://github.com/numba/numba/pull/1250#discussion_r32938067>:
>
> > @@ -148,6 +154,18 @@ def test_argmax_basic(self):
> >          npr, nbr = run_comparative(array_argmax, arr)
> >          self.assertPreciseEqual(npr, nbr)
> >
> > +    def test_median_odd(self):
> > +        arr = np.arange(101)
> > +        np.random.shuffle(arr)
> > +        npr, nbr = run_comparative(array_argmax, arr)
>
> Apparently you are not testing the right function :)
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/numba/numba/pull/1250/files#r32938067>.
>



-- 
Harsh

**Comment by seibert** [comment]
As a general question: How does Numpy implement median?

Also, numpy.median() for n-dimensional arrays finds the median over the flattened array.  I think we will want to do the same even if we don't support the axis argument.

File: numba/tests/test_array_reductions.py (hunk 1 above)
Comment (pitrou) [inline numba/tests/test_array_reductions.py]: No problem :)

Comment (hargup) [inline numba/tests/test_array_reductions.py]: :)

**Comment by hargup** [comment]
@seibert numpy uses the introselect [#1] algorithm, it is combination of the quickselect [#2] algorithm which gives a good average performance and the median of medians algorithm [#3] to handle the worst cases.

The implementation of introselection for numpy is written in C and is available here: https://github.com/numpy/numpy/blob/master/numpy/core/src/npysort/selection.c.src
Is there any way so this source file directly work for numba? If not probably we should spend energy on building such platform rather writing our own implementations for each algorithm, that can also save us a lot of effort in the long term.

[#1]: https://en.wikipedia.org/wiki/Introselect
[#2]: https://en.wikipedia.org/wiki/Quickselect
[#3]: https://en.wikipedia.org/wiki/Median_of_medians

**Comment by pitrou** [comment]
Well, Harsh's PR is the quickselect algorithm basically. We can just improve the pivot selection a bit (not necessarily a median of medians, but perhaps a median of three (start, end, middle)). Or we can randomize pivot selection as Harsh's TODO suggests.
(randomizing makes it more annoying to write tests for edge cases, though)

(writing this in C would make it unavailable to CUDA mode; not necessarily a deal-breaker but probably something we'd like to minimize)

**Comment by pitrou** [comment]
(rewriting my comment, sorry)

We can just improve the pivot selection a bit (not necessarily a median of medians, but perhaps a median of three (start, end, middle)). Or we can randomize pivot selection as the TODO suggests.
(randomizing makes it more annoying to write tests for edge cases, though)

Writing this in C would make it unavailable to CUDA mode; not necessarily a deal-breaker but probably something we'd like to minimize.

**Comment by seibert** [comment]
Given that access to random numbers on the GPU is a little tricky right now, randomized pivot selection would also prevent this from working on the GPU.

**Comment by hargup** [comment]
Since randomization is not a good option here, I'm using median of three as the pivot. I've also fixed the other mistakes pointed out by @pitrou

**Comment by pitrou** [comment]
The tests are failing :-) This is probably because you forgot to add typing information for the median method (and function). You can do so in typing/builtins.py and typing/npydecl.py.

File: numba/targets/arrayobj.py [hunk 8]
```diff
@@ -905,6 +905,58 @@ def array_argmax_impl(arry):
...
 
+@builtin
+@implement(numpy.median, types.Kind(types.Array))
+@implement("array.median", types.Kind(types.Array))
+def array_median(context, builder, sig, args):
+    def partition(A, low, high):
+        # median of three {low, middle, high}
+        middle = (low+high) // 2
+        if A[middle] > A[high]:
+            A[middle], A[high] = A[high], A[middle]
```
Comment (pitrou) [inline numba/targets/arrayobj.py]: There's no need to mutate the array for the median-of-three computation, you could simply use some local variables.

File: numba/targets/arrayobj.py [hunk 9]
```diff
@@ -905,6 +905,58 @@ def array_argmax_impl(arry):
...
+        if A[low] > A[high]:
+            A[high], A[low] = A[low], A[high]
+        if A[middle] > A[low]:
+            A[low], A[middle] = A[middle], A[low]
+
+        # choose the middle as the pivot
+        A[high], A[middle] = A[middle], A[high]
+
+        x = A[high]
+        i = low-1
```
Comment (pitrou) [inline numba/targets/arrayobj.py]: I still think you can start with `i = low` if you move `i += 1` after the swap inside the loop.

File: numba/tests/test_array_reductions.py [hunk 10]
```diff
@@ -148,6 +154,18 @@ def test_argmax_basic(self):
         npr, nbr = run_comparative(array_argmax, arr)
         self.assertPreciseEqual(npr, nbr)
 
+    def test_median_odd(self):
+        arr = np.arange(101)
+        np.random.shuffle(arr)
```
Comment (pitrou) [inline numba/tests/test_array_reductions.py]: In addition to the random case, I think it would be nice to also test with some special cases: a sorted array, a reverse-sorted array, an array with all equal values.

File: numba/tests/test_array_reductions.py [hunk 11]
```diff
@@ -77,6 +77,12 @@ def array_argmax(arr):
 def array_argmax_global(arr):
     return np.argmax(arr)
 
+def array_median(arr):
+    return arr.median()
+
+def array_median_global(arr):
```
Comment (pitrou) [inline numba/tests/test_array_reductions.py]: Note array_median_global isn't tested anywhere.

**Comment by hargup** [comment]
The tests are failing because the functions `partition` and `select` also needs to be compiled with numba [#1]. Another issue is that `array.flatten` is not implemented, I guess I'll be able to implement using `array.flat` which has already been implemented.

[#1]: https://travis-ci.org/numba/numba/jobs/68153752#L7579

**Comment by hargup** [comment]
@pitrou to compile the `partition` and `select` functions it looks like I'll have to do `context.compile(builder, partition, <sig>, <argument>)` is it right?

**Comment by hargup** [comment]
Well there is also `compile_only_no_cache` which doesn't require me pass the arguments but it is also fails with the error
```python
    from numba import typing

    def partition(A, low, high):
        # median of three {low, middle, high}
        middle = (low+high) // 2
        if A[middle] > A[high]:
            A[middle], A[high] = A[high], A[middle]
        if A[low] > A[high]:
            A[high], A[low] = A[low], A[high]
        if A[middle] > A[low]:
            A[low], A[middle] = A[middle], A[low]

        # choose the middle as the pivot
        A[high], A[middle] = A[middle], A[high]

        x = A[high]
        i = low-1
        for j in range(low, high):
            if A[j] <= x:
                i += 1
                A[i], A[j] = A[j], A[i]
        A[i+1], A[high] = A[high], A[i+1]
        return i + 1

    sig_partition = typing.signature(types.Kind(types.Number),
            (types.Kind(types.Array), types.Kind(types.Number), types.Kind(types.Number)))
    context.compile_only_no_cache(builder, partition, sig_partition)
```
```
Traceback (most recent call last):
  File "/home/hargup/210fs/numba/numba/lowering.py", line 170, in lower_block
    self.lower_inst(inst)
  File "/home/hargup/210fs/numba/numba/lowering.py", line 210, in lower_inst
    val = self.lower_assign(ty, inst)
  File "/home/hargup/210fs/numba/numba/lowering.py", line 359, in lower_assign
    return self.lower_expr(ty, value)
  File "/home/hargup/210fs/numba/numba/lowering.py", line 555, in lower_expr
    return self.lower_call(resty, expr)
  File "/home/hargup/210fs/numba/numba/lowering.py", line 526, in lower_call
    res = impl(self.builder, argvals)
  File "/home/hargup/210fs/numba/numba/targets/base.py", line 1058, in __call__
    return self._imp(self._context, builder, self._sig, args)
  File "/home/hargup/210fs/numba/numba/targets/arrayobj.py", line 938, in array_median
    context.compile_only_no_cache(builder, partition, sig_partition)
  File "/home/hargup/210fs/numba/numba/targets/base.py", line 855, in compile_only_no_cache
    locals=locals)
  File "/home/hargup/210fs/numba/numba/compiler.py", line 611, in compile_internal
    return pipeline.compile_extra(func)
  File "/home/hargup/210fs/numba/numba/compiler.py", line 317, in compile_extra
    return self.compile_bytecode(bc, func_attr=self.func_attr)
  File "/home/hargup/210fs/numba/numba/compiler.py", line 326, in compile_bytecode
    return self._compile_bytecode()
  File "/home/hargup/210fs/numba/numba/compiler.py", line 581, in _compile_bytecode
    return pm.run(self.status)
  File "/home/hargup/210fs/numba/numba/compiler.py", line 209, in run
    raise patched_exception
  File "/home/hargup/210fs/numba/numba/compiler.py", line 201, in run
    res = stage()
  File "/home/hargup/210fs/numba/numba/compiler.py", line 350, in stage_analyze_bytecode
    % (len(self.args), self.nargs))
TypeError: Failed at nopython (analyzing bytecode)
Signature mismatch: 1 argument types given, but function takes 3 arguments
```

**Comment by pitrou** [comment]
`typing.signature` takes a `*args`, not a tuple of args. Also, you have to pass actual type instances, not `types.Kind(something)`.

By the way, we can limit ourselves to 1d arrays if it makes things easier.

**Comment by hargup** [comment]
@pitrou please have a look at the new commits, I've generated the correct (hopefully) signatures, and compiled the code but the median code still cannot find the defintions of other function (partition and select). I've also tried adding function definitions to the cache but it doesn't help either.

**Comment by pitrou** [comment]
Yes, this is a bit trickier. See for an example `round_impl_binary()` in targets/builtins.py. Especially the `ExternalFunction` instantiation (note that it takes first a function name on the LLVM side, and second the function's Numba signature).

**Comment by hargup** [comment]
> See for an example round_impl_binary() in targets/builtins.py. Especially the ExternalFunction instantiation (note that it takes first a function name on the LLVM side, and second the function's Numba signature).

I tried that now at least numba recognizes that a function named `partition` and `select` exists for the `median` code to use. But I'm getting some error an `AssertionError` at the lowering stage [#1]. To investigate if it is an issue with my (algorithmic) implementation of median or if I'm using the `ExernalFunction` wrong, I tried to run the following code:

```python

@builtin
@implement(numpy.median, types.Kind(types.Array))
@implement("array.median", types.Kind(types.Array))
def array_median(context, builder, sig, args):
    from numba import typing, int64, float64

    def test_func(k):
        return k

    sig_test = typing.signature(types.int64, types.int64)
    cres = context.compile_only_no_cache(builder, test_func, sig_test)
    llvm_func_name = cres.fndesc.llvm_func_name
    _test = types.ExternalFunction(llvm_func_name, sig_test)

    def median(arry):
        x = _test(0)
        return arry[x]

    return context.compile_internal(builder, median, sig, args)
```
But this leads to a seg fault
```[1]    19042 segmentation fault (core dumped)  python median.py```


[#1]: https://travis-ci.org/numba/numba/jobs/68526388

**Comment by pitrou** [comment]
Have you tried to look at the generated LLVM code? You can use the NUMBA_DEBUG environment variable.

**Comment by pitrou** [comment]
Actually, I know what the problem is. ExternalFunction() works for functions with a "normal" calling convention, not for Numba-compiled functions. We'll have to find something else, it seems :/

**Comment by pitrou** [comment]
Ok, if you pull PR #1280 into your code, you should be able to call subroutines from your compiled Numba function.

**Comment by pitrou** [comment]
Basically something like:
```
    def test_func(k):
        return k

    sig_test = typing.signature(types.int64, types.int64)
    _test = context.compile_subroutine(builder, test_func, sig_test)
```

**Comment by hargup** [comment]
> Ok, if you pull PR #1280 into your code, you should be able to call subroutines from your compiled Numba function.

Thanks! :)

File: numba/targets/arrayobj.py (hunk 9 above)
Comment (hargup) [inline numba/targets/arrayobj.py]: Will also have to change the two lines after the loop:
```python
        A[i], A[high] = A[high], A[i]
        return i
```

File: numba/targets/arrayobj.py (hunk 8 above)
Comment (hargup) [inline numba/targets/arrayobj.py]: Any specific reason to not to mutate array for median-of-three computation? I see a way to use other local variables without decreasing the readability of the implemplementation.

File: numba/tests/test_array_reductions.py (hunk 11 above)
Comment (hargup) [inline numba/tests/test_array_reductions.py]: Thanks! It is not now tested.

File: numba/targets/arrayobj.py (hunk 8 above)
Comment (pitrou) [inline numba/targets/arrayobj.py]: I think it's more readable (and a bit more efficient). Also it's easier to reason about the algorithm if mutation is limited to a couple lines.

File: numba/tests/test_array_reductions.py (hunk 10 above)
Comment (hargup) [inline numba/tests/test_array_reductions.py]: I've added the tests you mentioned, please have a look.

Comment (pitrou) [inline numba/tests/test_array_reductions.py]: The tests look fine to me, thanks.

File: numba/targets/arrayobj.py [hunk 12]
```diff
@@ -904,6 +904,75 @@ def array_argmax_impl(arry):
     return context.compile_internal(builder, array_argmax_impl, sig, args)
 
 
+@builtin
+@implement(numpy.median, types.Kind(types.Array))
+def array_median(context, builder, sig, args):
+    from numba import typing, int64, float64
```
Comment (pitrou) [inline numba/targets/arrayobj.py]: Time for cosmetics :) int64 and float64 are already available as `types.int64` and `types.float64` (also, note that we tend to use `types.intp` for indices, rather than int64).

File: numba/targets/arrayobj.py [hunk 13]
```diff
@@ -904,6 +904,75 @@ def array_argmax_impl(arry):
...
+        while i != k:
+            if i < k:
+                low = i+1
+                i = _partition(temp_arry, low, high)
+            else:
+                high = i-1
+                i = _partition(temp_arry, low, high)
+        return temp_arry[k]
+
+    # XXX: Output is not necessarity float64, it should as the same type as that
```
Comment (pitrou) [inline numba/targets/arrayobj.py]: Indeed, so that would probably be `sig.args[0].dtype`.

**Comment by pitrou** [comment]
By the way, one more thing: you should add np.median as supported in the docs in `docs/source/reference/numpysupported.rst`.

File: docs/source/reference/numpysupported.rst [hunk 14]
```diff
@@ -158,6 +158,7 @@ The following top-level functions are supported:
 * :func:`numpy.round_`
 * :func:`numpy.zeros`
 * :func:`numpy.zeros_like`
+* :func:`numpy.median` (only the first argument)
```
Comment (pitrou) [inline docs/source/reference/numpysupported.rst]: Could you keep the list in sorted order? Otherwise, looks good to me.

Comment (hargup) [inline docs/source/reference/numpysupported.rst]: Done.

**Comment by pitrou** [comment]
Harsh, thank you! I'm merging the PR now.
